File: part-1/src/main/scala/controller/Simulator.scala
package controller

import akka.actor.typed.ActorRef
import conc.controller.Controller
import conc.model.monitor.{StartSync, StartSyncImpl, StopFlag, StopFlagImpl}
import conc.model.{Body, Boundary, P2d, V2d}
import view.{SimulationView}
import view.ViewActor.ViewCommands
import java.util
import java.util.{ArrayList, Random}
import scala.collection.mutable
import scala.collection.mutable.Seq
import scala.concurrent.Future

class Simulator(val nBodies: Int, val nSteps: Int, val nWorkers: Int):
  val bounds: Boundary = new Boundary(-4.0, -4.0, 4.0, 4.0)
  val rand = new Random(System.currentTimeMillis)
  val bodies= for i <- 0 until nBodies yield
      val x: Double = bounds.getX0 * 0.25 + rand.nextDouble * (bounds.getX1 - bounds.getX0) * 0.25
      val y: Double = bounds.getY0 * 0.25 + rand.nextDouble * (bounds.getY1 - bounds.getY0) * 0.25
      new Body(i, new P2d(x, y), new V2d(0, 0), 10)
  val list = bodies.toList
  val height = 620
  val width = 620

  def execute(): Unit =
    /** SETUP **/
   /* val sync: StartSync = new StartSyncImpl
    val flag: StopFlag = new StopFlagImpl
    val controller: Controller = Controller(sync, flag)
    val viewer: SimulationView = SimulationView(620, 620)
    viewer.registerListener(controller)
    var results = new Seq[Future[Body]]*/

    var vt = 0.0
    var dt = 0.001
//    var viewer = new SimulationView(utils, height, width)
//    var iter = 0
//    var bodyActor = BodyActor(bodies, b)
    
    /*for iter <- 0 until nSteps do 
      bodies.foreach(b => 
        var bodyActor = BodyActor(bodies, b, bounds, dt)
      )*/
    

@main def testMain: Unit =
  new Simulator(10, 2, 1)

File: part-1/src/main/java/conc/controller/Simulator.java
package conc.controller;

import conc.model.Body;
import conc.model.Boundary;
import conc.model.P2d;
import conc.model.V2d;
import conc.model.monitor.StartSync;
import conc.model.monitor.StartSyncImpl;
import conc.model.monitor.StopFlag;
import conc.model.monitor.StopFlagImpl;
import conc.view.SimulationView;

import java.util.ArrayList;
import java.util.List;
import java.util.Random;
import java.util.concurrent.*;

public class Simulator {
	/* bodies in the field */
	private ArrayList<Body> bodies;
	/* boundary of the field */
	private Boundary bounds;

	private final int nBodies, nSteps, nWorkers;

	public Simulator(int nBodies, int nSteps, int nWorkers) {
		this.nBodies = nBodies;
		this.nSteps = nSteps;
		this.nWorkers = nWorkers;
		testWithNumBodies();
	}
	
	public void execute() {
		/** SETUP **/
		StartSync sync = new StartSyncImpl();
		StopFlag flag = new StopFlagImpl();
		Controller controller = new Controller(sync, flag);
		SimulationView viewer = new SimulationView(620,620);
		viewer.registerListener(controller);

		/** EXECUTE **/
		var vt = 0.0;
		var dt = 0.001;
		var iter = 0;
		var exec = Executors.newFixedThreadPool(nWorkers);

		sync.waitStart();

		while(iter < nSteps){

			var results = new ArrayList<Future<Body>>();

			bodies.forEach(body -> {
				//var res = exec.submit(new ComputeAndUpdateVelocityTask(bodies, body, dt));
				//results.add(res);
			});

			var step = new ArrayList<Body>();
			results.forEach(bodyFuture -> {
				try {
					step.add(bodyFuture.get());
				} catch (Exception e) {
					e.printStackTrace();
				}
			});

			bodies = step;
			results.clear();

			bodies.forEach(body -> {
				//var res = exec.submit(new UpdatePosTask(body, dt));
				//results.add(res);
			});

			step.clear();
			results.forEach(bodyFuture -> {
				try {
					step.add(bodyFuture.get());
				} catch (Exception e) {
					e.printStackTrace();
				}
			});

			bodies = step;
			results.clear();

			bodies.forEach(body -> {
				//var res = exec.submit(new CheckBoundaryTask(body, bounds));
				//results.add(res);
			});

			step.clear();
			results.forEach(bodyFuture -> {
				try {
					step.add(bodyFuture.get());
				} catch (Exception e) {
					e.printStackTrace();
				}
			});

			bodies = step;
			results.clear();

			vt = vt + dt;
			iter++;

			if(!flag.isSet()){
				viewer.display(bodies, vt, iter, bounds);
			}
		}
	}

	private void testWithNumBodies() {
		bounds = new Boundary(-4.0, -4.0, 4.0, 4.0);

		Random rand = new Random(System.currentTimeMillis());
		bodies = new ArrayList<>();
		for (int i = 0; i < nBodies; i++) {
			double x = bounds.getX0()*0.25 + rand.nextDouble() * (bounds.getX1() - bounds.getX0()) * 0.25;
			double y = bounds.getY0()*0.25 + rand.nextDouble() * (bounds.getY1() - bounds.getY0()) * 0.25;
			Body b = new Body(i, new P2d(x, y), new V2d(0, 0), 10);
			bodies.add(b);
		}
	}
}
